# The wine dataset

[The wine dataset](https://archive.ics.uci.edu/dataset/109/wine) is the result of a chemical analysis of wines grown in the same region in Italy but derived from three different cultivars. The analysis determined the quantities of 13 constituents found in each of the three types of wines. It can be downloaded online, or using the following code:

```Python
from sklearn.datasets import load_wine
X, y = load_wine(return_X_y=True)
```

But you don't need to as the CSV is already stored in the files-folder.

## Data loading

Load the CSV into a numpy array. Show the first 5 rows of the array and print the shape. 
Values are shown in scientific notation.

In [1]:
import numpy as np

# Data Loading
# Using genfromtxt to skip the header row automatically
wine_dataset = np.genfromtxt('files/wine.csv', delimiter=',', skip_header=1)

print("First 5 records preview:")
print(wine_dataset[:5])
print(f"Shape: {wine_dataset.shape}")



First 5 records preview:
[[1.423e+01 1.710e+00 2.430e+00 1.560e+01 1.270e+02 2.800e+00 3.060e+00
  2.800e-01 2.290e+00 5.640e+00 1.040e+00 3.920e+00 1.065e+03 0.000e+00]
 [1.320e+01 1.780e+00 2.140e+00 1.120e+01 1.000e+02 2.650e+00 2.760e+00
  2.600e-01 1.280e+00 4.380e+00 1.050e+00 3.400e+00 1.050e+03 0.000e+00]
 [1.316e+01 2.360e+00 2.670e+00 1.860e+01 1.010e+02 2.800e+00 3.240e+00
  3.000e-01 2.810e+00 5.680e+00 1.030e+00 3.170e+00 1.185e+03 0.000e+00]
 [1.437e+01 1.950e+00 2.500e+00 1.680e+01 1.130e+02 3.850e+00 3.490e+00
  2.400e-01 2.180e+00 7.800e+00 8.600e-01 3.450e+00 1.480e+03 0.000e+00]
 [1.324e+01 2.590e+00 2.870e+00 2.100e+01 1.180e+02 2.800e+00 2.690e+00
  3.900e-01 1.820e+00 4.320e+00 1.040e+00 2.930e+00 7.350e+02 0.000e+00]]
Shape: (178, 14)


The shape is 178 rows by 14 columns. This means we have 178 different types of wine in our database. 


Explore the data by showing the average value of every column.

If you see this, the means are shown in scientific notation:

![](files/2025-09-17-20-01-43.png)

It's technically correct but rather difficult to interprete. To show these values as numbers, you can use the np-method set_printoptions().



In [2]:
# Adjust print options to avoid scientific notation for readability
np.set_printoptions(precision=4, suppress=True)

col_means = np.mean(wine_dataset, axis=0)
print("\nColumn Means (Readable):")
print(col_means)




Column Means (Readable):
[ 13.0006   2.3363   2.3665  19.4949  99.7416   2.2951   2.0293   0.3619
   1.5909   5.0581   0.9574   2.6117 746.8933   0.9382]


## Normalization

You will notice that some columns have a different value range. The second-to-last column has an average of 746, while other columns have an average of less than 1.  In later courses, you will learn that some Machine Learning algorithms (k-Means clustering or logistic regression) really struggle with this and force you to normalise the data. 

Solution is to rescale numerical data to a standard range to ensure all features in a dataset have the same scale. So, one feature with a large range (e.g., salary) doesn't unfairly dominate another with a smaller range (e.g., years of experience). This process of rescaling is called normalization.

When normalizing we recalculate all these values to be between for example 0 and 1. In machine learning you'd normally use a scaler, like sklearn.preprocessing.StandardScaler, but all this scaler does is apply some NumPy-magic to the columns.   
Let's start with an example:

In [ ]:
X = np.array([
    [ 1, 200,  3],
    [ 2, 180,  6],
    [ 3, 160,  9],
    [ 4, 220, 12],
    [ 5, 240, 15]
], dtype=float)

**Try Min–Max Scaling (to [0,1])**

Rescale each column so its minimum = 0 and maximum = 1.

Tasks:
* Compute the column-wise minimum and maximum with np.min / np.max.
* Apply the formula:

![](files/2025-09-17-20-22-45.png)

* Verify that every column is now between 0 and 1.

In [4]:
# Normalization Section
# Example Data
sample_X = np.array([
    [ 1, 200,  3],
    [ 2, 180,  6],
    [ 3, 160,  9],
    [ 4, 220, 12],
    [ 5, 240, 15]
], dtype=float)

min_vals = np.min(sample_X, axis=0)
max_vals = np.max(sample_X, axis=0)
scaled_minmax = (sample_X - min_vals) / (max_vals - min_vals)
print("\nMin-Max Scaled Data:")
print(scaled_minmax)




Min-Max Scaled Data:
[[0.   0.5  0.  ]
 [0.25 0.25 0.25]
 [0.5  0.   0.5 ]
 [0.75 0.75 0.75]
 [1.   1.   1.  ]]


**Try Z-Score Standardization (mean = 0, std = 1)**

In stead of min-max scaling you could apply a Z-score standarization. Make each column have mean 0 and standard deviation 1. This means the numbers won't be between 0 and 1 but between -1 and +1.

Tasks:
* Compute the column-wise mean and standard deviation (np.mean, np.std).
* Apply the formula:

![](files/2025-09-17-20-27-40.png)

* Check results:
	* Column means should be very close to 0.
	* Column standard deviation should be 1.

In [5]:
# Method 2: Z-Score Standardization (Mean=0, Std=1)
mean_vals = np.mean(sample_X, axis=0)
std_vals = np.std(sample_X, axis=0)
scaled_zscore = (sample_X - mean_vals) / std_vals

print("\nZ-Score Standardized Data:")
print(scaled_zscore)
print(f"Verify Means: {np.mean(scaled_zscore, axis=0)}")
print(f"Verify Stds: {np.std(scaled_zscore, axis=0)}")




Z-Score Standardized Data:
[[-1.4142  0.     -1.4142]
 [-0.7071 -0.7071 -0.7071]
 [ 0.     -1.4142  0.    ]
 [ 0.7071  0.7071  0.7071]
 [ 1.4142  1.4142  1.4142]]
Verify Means: [ 0.  0. -0.]
Verify Stds: [1. 1. 1.]


Which of the two methods above is most sensitive to outliers? (This is a theoretical question, so in the next code-block you should only enter text.)

In [6]:
# Theoretical Question Answer
outlier_sensitivity_answer = """
While both methods react to outliers, Min-Max scaling is drastically more sensitive. 
Because it compresses all data into a fixed [0, 1] window based on the absolute extreme values, 
a single outlier pushes the 'max' far out, squashing all normal data points very close to 0. 
Z-score spreads data based on standard deviation; while the std increases with an outlier, 
the relative positioning of normal data remains more intact compared to the hard compression of Min-Max.
"""
print("\nOutlier Sensitivity Analysis:")
print(outlier_sensitivity_answer)



Outlier Sensitivity Analysis:

While both methods react to outliers, Min-Max scaling is drastically more sensitive. 
Because it compresses all data into a fixed [0, 1] window based on the absolute extreme values, 
a single outlier pushes the 'max' far out, squashing all normal data points very close to 0. 
Z-score spreads data based on standard deviation; while the std increases with an outlier, 
the relative positioning of normal data remains more intact compared to the hard compression of Min-Max.



**Apply Min–Max Scaling (to [0,1]) to the wine dataset**   

Normalize the wine dataset as well!

In [7]:
# Apply MinMax to Wine Dataset
wine_min = np.min(wine_dataset, axis=0)
wine_max = np.max(wine_dataset, axis=0)
wine_normalized = (wine_dataset - wine_min) / (wine_max - wine_min)

print("\nWine Data Normalized (First 5 rows):")
print(wine_normalized[:5])




Wine Data Normalized (First 5 rows):
[[0.8421 0.1917 0.5722 0.2577 0.6196 0.6276 0.5738 0.283  0.5931 0.372
  0.4553 0.9707 0.5613 0.    ]
 [0.5711 0.2055 0.4171 0.0309 0.3261 0.5759 0.5105 0.2453 0.2744 0.2645
  0.4634 0.7802 0.5506 0.    ]
 [0.5605 0.3202 0.7005 0.4124 0.337  0.6276 0.6118 0.3208 0.7571 0.3754
  0.4472 0.696  0.6469 0.    ]
 [0.8789 0.2391 0.6096 0.3196 0.4674 0.9897 0.6646 0.2075 0.5584 0.5563
  0.3089 0.7985 0.8573 0.    ]
 [0.5816 0.3656 0.8075 0.5361 0.5217 0.6276 0.4958 0.4906 0.4448 0.2594
  0.4553 0.6081 0.326  0.    ]]


You may feel a sting now, a nagging thought in the back of your mind: We've drastically changed the numbers. How can we do predictions on this. Suppose we have another wine with the following values:

```
13.71,1.86,2.36,16.6,101.0,2.61,2.88,0.27,1.69,3.8,1.11,4.0,1035.0,0
```

Could we transform it into a normalized wine?

(Remember that we still have the datarow with minimum values and with maximum values.)

In [8]:
# Transforming a new observation
new_wine_sample = np.array([13.71, 1.86, 2.36, 16.6, 101.0, 2.61, 2.88, 0.27, 1.69, 3.8, 1.11, 4.0, 1035.0, 0])
new_wine_norm = (new_wine_sample - wine_min) / (wine_max - wine_min)
print("\nNormalized New Sample:")
print(new_wine_norm)


Normalized New Sample:
[0.7053 0.2213 0.5348 0.3093 0.337  0.5621 0.5359 0.2642 0.4038 0.215
 0.5122 1.     0.5399 0.    ]


Going back is also important. Can you recalculate the scaled row into it's original values?

In [9]:
# Inverse Transform (Denormalize)
reconstructed = (new_wine_norm * (wine_max - wine_min)) + wine_min
print("\nReconstructed Original Values:")
print(reconstructed)




Reconstructed Original Values:
[  13.71    1.86    2.36   16.6   101.      2.61    2.88    0.27    1.69
    3.8     1.11    4.   1035.      0.  ]


## Covariance and correlation

Covariance and correlation are measures on how the data in distinct columns moves in a similar or opposite way. 

Covariance is a statistical which measures the relationship between a pair of random variables where a change in one variable causes a change in another variable. It assesses how much two variables change together from their mean values. 

Using Numpy, you can display a covariance matrix to show the covariance between the columns.

In [10]:
# Covariance & Correlation
cov_matrix = np.cov(wine_dataset, rowvar=False)
corr_matrix = np.corrcoef(wine_dataset, rowvar=False)

print("\nCovariance Matrix (Snippet - Top Left 3x3):")
print(cov_matrix[:3, :3])

print("\nCorrelation Matrix (Snippet - Top Left 3x3):")
print(corr_matrix[:3, :3])


Covariance Matrix (Snippet - Top Left 3x3):
[[0.6591 0.0856 0.0471]
 [0.0856 1.248  0.0503]
 [0.0471 0.0503 0.0753]]

Correlation Matrix (Snippet - Top Left 3x3):
[[1.     0.0944 0.2115]
 [0.0944 1.     0.164 ]
 [0.2115 0.164  1.    ]]


What you get is a 14 by 14 matrix of values. The first row looks like this:

![](files/2025-09-17-21-16-33.png)

These are the covariances between the first column and each of the 14  columns. You notice two main things:
- Some numbers are positive and some numbers are negative
- Some numbers are big, other really small

If a number is negative it means that if one goes up the other goes down. If it is positive both tend to move in the same direction. That is usefull information.

The size of the number means nothing at all. We are comparing data on different chemicals to each other, which is as good as comparing apples and oranges. Another analogy would be saying somebody got a "7" on a test. Is that good or bad? No way of knowing if you don't know the maximum score for that test. That is why covariance is not used much. Covariance helps us understand the direction of the relationship but not how strong it is because the calculated number depends on the units used. The covariance matrix just helps you to see how two columns are connected.

If you need a measure of the strength and direction of the linear relationship between two columns, you calculate the correlation matrix. It is derived from covariance, but standardized and ranges between -1 and 1. Unlike covariance, which only indicates the direction of the relationship, correlation provides a standardized measure that you can use to detect the strength of the relationship between 2 variables.

In [13]:
# Calculate the correlation matrix
corr_matrix = np.corrcoef(wine_dataset, rowvar=False)
print(corr_matrix)

[[ 1.      0.0944  0.2115 -0.3102  0.2708  0.2891  0.2368 -0.1559  0.1367
   0.5464 -0.0717  0.0723  0.6437 -0.3282]
 [ 0.0944  1.      0.164   0.2885 -0.0546 -0.3352 -0.411   0.293  -0.2207
   0.249  -0.5613 -0.3687 -0.192   0.4378]
 [ 0.2115  0.164   1.      0.4434  0.2866  0.129   0.1151  0.1862  0.0097
   0.2589 -0.0747  0.0039  0.2236 -0.0496]
 [-0.3102  0.2885  0.4434  1.     -0.0833 -0.3211 -0.3514  0.3619 -0.1973
   0.0187 -0.274  -0.2768 -0.4406  0.5179]
 [ 0.2708 -0.0546  0.2866 -0.0833  1.      0.2144  0.1958 -0.2563  0.2364
   0.2     0.0554  0.066   0.3934 -0.2092]
 [ 0.2891 -0.3352  0.129  -0.3211  0.2144  1.      0.8646 -0.4499  0.6124
  -0.0551  0.4337  0.6999  0.4981 -0.7192]
 [ 0.2368 -0.411   0.1151 -0.3514  0.1958  0.8646  1.     -0.5379  0.6527
  -0.1724  0.5435  0.7872  0.4942 -0.8475]
 [-0.1559  0.293   0.1862  0.3619 -0.2563 -0.4499 -0.5379  1.     -0.3658
   0.1391 -0.2626 -0.5033 -0.3114  0.4891]
 [ 0.1367 -0.2207  0.0097 -0.1973  0.2364  0.6124  0.6527 -0.365

The same row now looks like this:

![](files/2025-09-17-21-20-21.png)

Again, we see different values:
- The first value is 1. This is when comparing the column to itself, which is a perfect correlation.
- Negative values: one goes up, the other goes down
- Positive values: the values move in the same direction

But mainly interesting is the size of the numbers. Explain what a correlation value of -1, +1, 0, and all between means. (Code block text.)

In [11]:
# Interpretation Answer
correlation_interpretation = """
Correlation Coefficient Interpretation:
+1.0 : Perfect Positive Linear Relationship (As X goes up, Y goes up exactly proportionally).
-1.0 : Perfect Negative Linear Relationship (As X goes up, Y goes down exactly proportionally).
 0.0 : No Linear Relationship (Changes in X do not predict changes in Y linearly).
Values between these extremes indicate the strength. 
E.g., 0.8 is a strong positive link, while -0.3 is a weak negative link.
Unlike covariance, these values are unitless and comparable across different features.
"""
print("\nCorrelation Interpretation:")
print(correlation_interpretation)




Correlation Interpretation:

Correlation Coefficient Interpretation:
+1.0 : Perfect Positive Linear Relationship (As X goes up, Y goes up exactly proportionally).
-1.0 : Perfect Negative Linear Relationship (As X goes up, Y goes down exactly proportionally).
 0.0 : No Linear Relationship (Changes in X do not predict changes in Y linearly).
Values between these extremes indicate the strength. 
E.g., 0.8 is a strong positive link, while -0.3 is a weak negative link.
Unlike covariance, these values are unitless and comparable across different features.

